test/test_special_elixirs.jl

```diff
@@ -22,6 +22,9 @@ const EXAMPLES_DIR = joinpath(pathof(Trixi) |> dirname |> dirname, "examples")
     mean_eoc = convergence_test(@__MODULE__, joinpath(EXAMPLES_DIR, "2d", "elixir_advection_extended_curved.jl"), 3)
     @test isapprox(mean_eoc[:l2], [4.0], rtol=0.01)
 
+    mean_eoc = convergence_test(@__MODULE__, joinpath(EXAMPLES_DIR, "2d", "elixir_euler_nonperiodic_p4est.jl"), 3)
+    @test isapprox(mean_eoc[:l2], [3.54, 3.50, 3.50, 3.52], rtol=0.01)
```

Comment:
@efaulhaber Can you tell me agian why the EOCs are so low?

Reply:
I guess that's because it's using an unstructured mesh, but I'm not an expert.
It's this one:
![grafik](https://user-images.githubusercontent.com/44124897/119278907-a9e4f400-bc28-11eb-8814-82f5a7c07ba2.png)

Reply:
@andrewwinters5000 could we get your take on this? Such a mesh with straight sides should give us the full EOC, shouldn't it?

Reply:
There are lots of kite-like elements (when one of the angles is very obtuse and the element is approaching a triangle) in this mesh which can mess with the quality of the EOCs.

Reply:
How does the EOC look in the last few refinement levels? Sometimes the mean EOC can get polluted with sub-optimal values from too coarse resolution

Reply:
```
l2
rho                 rho_v1              rho_v2              rho_e
error     EOC       error     EOC       error     EOC       error     EOC
5.69e-03  -         4.52e-03  -         4.52e-03  -         9.97e-03  -
5.81e-04  3.29      4.78e-04  3.24      4.78e-04  3.24      1.03e-03  3.27
4.20e-05  3.79      3.52e-05  3.76      3.52e-05  3.76      7.54e-05  3.78
2.86e-06  3.88      2.41e-06  3.87      2.41e-06  3.87      5.06e-06  3.90

mean      3.65      mean      3.62      mean      3.62      mean      3.65
----------------------------------------------------------------------------------------------------
linf
rho                 rho_v1              rho_v2              rho_e
error     EOC       error     EOC       error     EOC       error     EOC
5.13e-02  -         5.34e-02  -         5.34e-02  -         9.03e-02  -
4.55e-03  3.49      3.58e-03  3.90      3.58e-03  3.90      6.91e-03  3.71
2.86e-04  3.99      2.98e-04  3.58      2.98e-04  3.58      5.39e-04  3.68
4.39e-05  2.70      5.32e-05  2.49      5.32e-05  2.49      7.94e-05  2.76

mean      3.40      mean      3.32      mean      3.32      mean      3.38
----------------------------------------------------------------------------------------------------
```

Reply:
It could also be related to the troubles I am encountering (and still trying to solve) for EOCs on the warped mesh. Where I get similar mean EOC values that approach `polydeg+1/2` instead of `polydeg+1`

Reply:
This mesh here produces a bit better results:

![grafik](https://user-images.githubusercontent.com/44124897/121504928-abf9d180-c9e2-11eb-9a17-18979f99c9f2.png)

```
l2
rho                 rho_v1              rho_v2              rho_e
error     EOC       error     EOC       error     EOC       error     EOC
1.53e-03  -         1.20e-03  -         1.20e-03  -         2.65e-03  -
1.07e-04  3.84      8.47e-05  3.83      8.47e-05  3.83      1.84e-04  3.85
8.36e-06  3.67      6.79e-06  3.64      6.79e-06  3.64      1.47e-05  3.65

mean      3.76      mean      3.73      mean      3.73      mean      3.75
----------------------------------------------------------------------------------------------------
linf
rho                 rho_v1              rho_v2              rho_e
error     EOC       error     EOC       error     EOC       error     EOC
1.98e-02  -         8.63e-03  -         8.63e-03  -         1.51e-02  -
1.12e-03  4.15      5.79e-04  3.90      5.79e-04  3.90      1.18e-03  3.68
1.37e-04  3.03      6.12e-05  3.24      6.13e-05  3.24      1.49e-04  2.98

mean      3.59      mean      3.57      mean      3.57      mean      3.33
----------------------------------------------------------------------------------------------------
```

Reply:
Hmm, what `cfl` did you use? Is it taken small enough to ensure that the spatial errors dominate the approximation?

Reply:
This is all done with `cfl=1.0`. Here are the results for both meshes with `cfl=0.1`.
```
l2
rho                 rho_v1              rho_v2              rho_e
error     EOC       error     EOC       error     EOC       error     EOC
5.69e-03  -         4.52e-03  -         4.52e-03  -         9.97e-03  -
5.81e-04  3.29      4.78e-04  3.24      4.78e-04  3.24      1.03e-03  3.27
4.20e-05  3.79      3.52e-05  3.76      3.52e-05  3.76      7.54e-05  3.78

mean      3.54      mean      3.50      mean      3.50      mean      3.52
----------------------------------------------------------------------------------------------------
linf
rho                 rho_v1              rho_v2              rho_e
error     EOC       error     EOC       error     EOC       error     EOC
5.13e-02  -         5.34e-02  -         5.34e-02  -         9.03e-02  -
4.55e-03  3.49      3.58e-03  3.90      3.58e-03  3.90      6.91e-03  3.71
2.86e-04  3.99      2.98e-04  3.58      2.98e-04  3.58      5.39e-04  3.68

mean      3.74      mean      3.74      mean      3.74      mean      3.69
----------------------------------------------------------------------------------------------------
```
```
l2
rho                 rho_v1              rho_v2              rho_e
error     EOC       error     EOC       error     EOC       error     EOC
1.53e-03  -         1.20e-03  -         1.20e-03  -         2.65e-03  -
1.07e-04  3.84      8.47e-05  3.83      8.47e-05  3.83      1.84e-04  3.85
8.36e-06  3.67      6.79e-06  3.64      6.79e-06  3.64      1.47e-05  3.65

mean      3.76      mean      3.73      mean      3.73      mean      3.75
----------------------------------------------------------------------------------------------------
linf
rho                 rho_v1              rho_v2              rho_e
error     EOC       error     EOC       error     EOC       error     EOC
1.98e-02  -         8.63e-03  -         8.63e-03  -         1.51e-02  -
1.12e-03  4.15      5.79e-04  3.90      5.79e-04  3.90      1.18e-03  3.68
1.37e-04  3.03      6.12e-05  3.24      6.13e-05  3.24      1.49e-04  2.98

mean      3.59      mean      3.57      mean      3.57      mean      3.33
----------------------------------------------------------------------------------------------------
```
It doesn't change anything for the second mesh. With the first mesh, the L2 errors become worse while the Linf errors become better.

Reply:
Interesting...what do you think @gregorgassner ? I am out of ideas off the top of my head

Reply:
This is a long shot, but It could also be that in `convergence_test` when we divide by `log(1/2)` is not 100% kosher anymore. To be safe we could try to divide by log( DOFs fine / DOFs coarse)

Reply:
This is all uniformly refined.

Reply:
Okay, I wasn't sure if there was any shenanigans with local refinement

Reply:
@efaulhaber Are the meshes used here curved? Andrew noted some weirdness regarding the EOC when using the volume mapping approach instead of the transfinite mapping from the boundaries. But if it is all straight-sided, it shouldn't make a difference I think.

Also, do you have another mesh for which you have tested the EOC with the P4estMesh?

Reply:
Right, for straight-sided elements the metric terms, Jacobian, etc. should all be okay. If you set Dirichlet boundary conditions with the exact solution instead of periodic does it change anything?

Reply:
Both meshes are uncurved.
This is all non-periodic using the exact Dirichlet boundary conditions (I can't build a periodic unstructured mesh from a mesh file).
Here's the same convergence test using a non-periodic *structured* `P4estMesh` (with 5 x 5 trees):
```
l2
rho                 rho_v1              rho_v2              rho_e
error     EOC       error     EOC       error     EOC       error     EOC
3.43e-04  -         3.00e-04  -         3.00e-04  -         7.64e-04  -
1.55e-05  4.47      1.47e-05  4.35      1.47e-05  4.35      3.92e-05  4.29
9.34e-07  4.05      8.91e-07  4.04      8.91e-07  4.04      2.42e-06  4.02

mean      4.26      mean      4.20      mean      4.20      mean      4.15
----------------------------------------------------------------------------------------------------
linf
rho                 rho_v1              rho_v2              rho_e
error     EOC       error     EOC       error     EOC       error     EOC
2.61e-03  -         2.21e-03  -         2.21e-03  -         5.79e-03  -
8.22e-05  4.99      1.00e-04  4.46      1.00e-04  4.46      3.87e-04  3.90
6.54e-06  3.65      6.83e-06  3.87      6.83e-06  3.87      2.33e-05  4.06

mean      4.32      mean      4.17      mean      4.17      mean      3.98
----------------------------------------------------------------------------------------------------
```

Reply:
Ah, ok. It could be the boundary conditions then. For perfect EOCs you usually need periodic boundaries, so this might screwing with things. Although... did you run your EOC investigations with periodic boundaries or Dirichlet, @andrewwinters5000?

Is this example above the for the waving flag? Or the Jesse mesh?

Reply:
All these examples are the uncurved square `[-1, 1]^2`. The convergence test above of the structured mesh still uses Dirichlet boundaries.

Reply:
Hm. What's the EOC on the waving flag? That's the most complex example we can do right now that has periodic boundaries, isn't it?

Reply:
All of my investigations have been using periodic BCS. The straight-sided metric are okay because you get the correct convergence on a structured version of the same mesh. It could be that setting Dirichlet BCs needs a correction term in each RK stage due to the time dependent boundary condition

Reply:
@sloede Waving flag looks like this:
```
l2
scalar
error     EOC
1.06e-02  -
5.61e-04  4.23
3.38e-05  4.05
2.11e-06  4.00

mean      4.10
----------------------------------------------------------------------------------------------------
linf
scalar
error     EOC
5.86e-02  -
4.73e-03  3.63
3.37e-04  3.81
2.14e-05  3.98

mean      3.81
----------------------------------------------------------------------------------------------------
```
But again, that's structured. The structured example using Dirichlet boundaries above worked as well. I can't do unstructured with periodic boundaries.

Reply:
OK, then I vote to create an issue calling for testing EOC for unstructured P4est once unstructured + periodic works. With the issue created, this conversation can be closed I think.

Thanks for being thorough and doing (once more) all the EOC checks, @efaulhaber!

Reply:
Something is fishy here...does this use `calc_interface_flux!` from `dg_unstructured` or did you write a new one?

Reply:
I wrote a new one, but it's very similar to yours.

Reply:
Okay, I am still trying to track down the EOCs issue in my implementation (so it could have a bug in it). I was thinking that the problem was the mapping, but our EOCs seem to indicate that it is also a problem for straight-sided elements

Reply:
Can you convert my [`.inp` mesh file](https://gist.github.com/efaulhaber/63ff2ea224409e55ee8423b3a33e316a) to one that you can read somehow? Then you could use the same mesh with your solver (maybe even make it periodic?).

Reply:
[This](https://gist.github.com/efaulhaber/a075f8ec39a67fa9fad8f6f84342cbca) is the second (less crazy) mesh.